File: weird_captcha_gym/shared_runtime/app/mechanics/comparator_engine.js
(() => {
  'use strict';
  const esc = s => String(s).replaceAll('&','&amp;').replaceAll('<','&lt;').replaceAll('"','&quot;');
  async function render(state, helpers) {
    document.body.dataset.mechanic = 'comparator-engine';
    const manual = state.manual_readings === true;
    let measured = !manual;
    const full = state.control_condition?.interaction === 'full';
    const row = state.slides.map(s => s.id), slides = Object.fromEntries(state.slides.map(s => [s.id,s]));
    let cursor=0, advances=0, exchanges=0, readings=manual?0:1, terminal=null, busy=false, pending=null;
    const events=[];
    const actions = manual?['advance','exchange','weigh','seal']:['advance','exchange','seal'];
    const controls = actions.map(a => full
      ? `<div class="ce-control"><div class="ce-lever" data-action="${a}" aria-label="${a} lever"><div class="ce-track"></div><span class="ce-knob"></span><i>↓</i></div><b>${a.toUpperCase()}</b></div>`
      : `<button class="ce-button" data-action="${a}"><b>${a.toUpperCase()}</b></button>`).join('');
    helpers.app.innerHTML = `<section class="ce-engine" data-challenge-id="${esc(state.challenge_id)}" data-interaction="${full?'full':'simplified'}">
      <header class="ce-header"><div><span>THE ENTOMOLOGICAL WEIGHTS OFFICE / No. 09</span><h1>The Comparator Engine</h1></div><div class="ce-seal">STANDARD<br>OF MASS<br><b>⚖</b></div></header>
      <div class="ce-rule">LIGHTEST <span>← &nbsp; Seal the entire frame in ascending weight &nbsp; →</span> HEAVIEST</div>
      <main class="ce-instrument"><div class="ce-rail"><div class="ce-carriage"><div class="ce-lamp"><span>HEAVIER SIDE</span><svg viewBox="0 0 140 58"><path d="M70 22 L55 52 H85 Z" fill="#bc9655"/><g class="ce-beam"><path d="M15 22 H125" stroke="#f8df9a" stroke-width="5"/><circle class="ce-left" cx="18" cy="22" r="10"/><circle class="ce-right" cx="122" cy="22" r="10"/></g></svg></div><div class="ce-windows"><i></i><i></i></div></div><div class="ce-row"></div></div>
      <div class="ce-legend"><span aria-hidden="true"></span><b class="ce-pair-label"></b></div></main>
      <section class="ce-console"><aside><span>MECHANICAL RESERVE</span><div class="ce-counter"></div><p class="ce-pass">PASS 1</p></aside><div class="ce-controls">${controls}</div><aside class="ce-console-spacer" aria-hidden="true"></aside></section>
      <footer><div class="readout" data-status="idle">FRAME OPEN</div></footer>
      <div class="ce-verdict" hidden></div></section>`;
    const root=helpers.app.querySelector('.ce-engine');
    if(manual)root.classList.add('ce-metered');
    const drawMoth = (canvas,s) => {
      const c=canvas.getContext('2d'); c.clearRect(0,0,80,118); c.save();c.translate(40,59);
      const scale=.56+s.size_band*.095; c.scale(scale,scale);c.strokeStyle='#563f23';c.fillStyle='#b7a074';c.lineWidth=1.6;
      for(const side of [-1,1]) {c.save();c.scale(side,1);c.beginPath();c.moveTo(0,-16);c.bezierCurveTo(18,-57,47,-48,37,-9);c.bezierCurveTo(62,23,30,50,4,19);c.closePath();c.fill();c.stroke();
        for(let j=0;j<4;j++){c.beginPath();c.moveTo(0,-10+j*7);c.lineTo(33,-32+j*16);c.stroke();}
        c.fillStyle='#4a3929'; for(let j=0;j<s.spots;j++){c.beginPath();c.ellipse(18+j*5,-20+j*15,3+s.engraving%3,5,0,0,Math.PI*2);c.fill();}c.restore();}
      c.fillStyle='#433222';c.beginPath();c.ellipse(0,0,4,30,0,0,Math.PI*2);c.fill();c.beginPath();c.moveTo(0,-27);c.lineTo(-12,-43);c.moveTo(0,-27);c.lineTo(12,-43);c.stroke();c.restore();
    };
    function draw() {
      root.style.setProperty('--n',row.length);
      root.querySelector('.ce-row').innerHTML=row.map((id,i)=>`<article class="ce-slide ${i===cursor||i===cursor+1?'is-current':''}"><span class="ce-slot">${String(i+1).padStart(2,'0')}</span><canvas width="80" height="118"></canvas><b>${esc(id)}</b>${manual?`<span class="ce-band">SIZE ${slides[id].size_band+1}</span>`:''}</article>`).join('');
      root.querySelectorAll('.ce-slide canvas').forEach((c,i)=>drawMoth(c,slides[row[i]]));
      root.querySelector('.ce-carriage').style.left=`${cursor*100/row.length}%`;
      root.querySelector('.ce-carriage').style.width=`${200/row.length}%`;
      const heavier=measured?(state.runtime_weights[row[cursor]]>state.runtime_weights[row[cursor+1]]?'left':'right'):null;
      root.querySelector('.ce-beam').setAttribute('transform',`rotate(${heavier===null?0:heavier==='left'?-12:12} 70 22)`);
      root.querySelector('.ce-left').setAttribute('fill',heavier==='left'?'#fff3ac':'#645136');
      root.querySelector('.ce-right').setAttribute('fill',heavier==='right'?'#fff3ac':'#645136');
      root.querySelector('.ce-pair-label').textContent=`${row[cursor]} · ${row[cursor+1]} / ${heavier?heavier.toUpperCase()+' HEAVIER':'NOT WEIGHED'}`;
      root.querySelector('.ce-counter').innerHTML=`<b>${state.limits.readings-readings}</b> readings left <br><b>${state.limits.levers-advances-exchanges}</b> ${manual?'carriage moves':'lever pulls'} left`;
      root.querySelector('.ce-pass').textContent=`PASS ${Math.floor(advances/(row.length-1))+1} · ${advances%(row.length-1)===0 && advances?'↻ BELL / WRAPPED':'PAIR '+(cursor+1)+'–'+(cursor+2)}`;
    }
    function show(title,button,callback) {
      const box=root.querySelector('.ce-verdict');box.hidden=false;
      box.innerHTML=`<strong>${esc(title)}</strong>${button?`<button>${esc(button)}</button>`:''}`;
      if(button)box.querySelector('button').onclick=callback;
    }
    async function submit() {
      if(busy)return;busy=true;
      show('CHECKING SEAL');
      try {
        const response=await fetch('/result',{method:'POST',headers:{'content-type':'application/json'},body:JSON.stringify(pending)});
        if(!response.ok)throw new Error('Server unavailable');
        const result=await response.json();
        if(result.passed===true){helpers.setReadout('PASS','passed');show('PASS');}
        else if(result.passed===false && result.state){helpers.setReadout('FAIL','error');show('FAIL','LOAD FRESH FRAME',()=>helpers.render(result.state));}
        else throw new Error(result.error||'No grade received');
      } catch(error){busy=false;show('SEAL NOT RECEIVED','RETRY SEAL',submit);}
    }
    function act(type,gesture) {
      if(terminal||busy)return;
      events.push({seq:events.length+1,type,cursor,pair:row.slice(cursor,cursor+2),input_source:full?'lever_drag':'button',...(gesture?{gesture}:{})});
      if(type==='seal')terminal=row.every((a,i)=>!i||state.runtime_weights[row[i-1]]<state.runtime_weights[a])?'sorted':'unsorted_seal';
      else if(type==='weigh'){if(readings>=state.limits.readings)terminal='comparison_exhausted';else{readings++;measured=true;}}
      else if(advances+exchanges>=state.limits.levers)terminal='lever_exhausted';
      else if(type==='advance'&&!manual&&readings>=state.limits.readings)terminal='comparison_exhausted';
      else if(type==='advance'){cursor=(cursor+1)%(row.length-1);advances++;if(manual)measured=false;else readings++;}
      else{[row[cursor],row[cursor+1]]=[row[cursor+1],row[cursor]];exchanges++;}
      draw();
      if(terminal){pending={mechanic_id:state.mechanic_id,task_id:state.task_id,challenge_id:state.challenge_id,control_condition:state.control_condition||null,events,final_order:[...row],counts:{advances,exchanges,readings},completed:terminal==='sorted'};submit();}
      else helpers.setReadout(type==='exchange'?'PAIR EXCHANGED':type==='weigh'?'PAIR WEIGHED':'CARRIAGE ADVANCED','idle');
    }
    root.querySelectorAll('.ce-button').forEach(b=>b.addEventListener('click',()=>act(b.dataset.action)));
    root.querySelectorAll('.ce-lever').forEach(lever=>{
      let drag=null;
      const point=e=>{const b=lever.getBoundingClientRect();return {x:(e.clientX-b.left)/b.width,y:(e.clientY-b.top)/b.height};};
      lever.addEventListener('pointerdown',e=>{
        if(e.button!==0||terminal||busy||!e.target.classList.contains('ce-knob'))return;
        const p=point(e);if(p.x<.25||p.x>.75||p.y<0||p.y>.4)return;
        drag={id:e.pointerId,...p};lever.setPointerCapture(e.pointerId);e.preventDefault();
      });
      lever.addEventListener('pointermove',e=>{if(drag&&drag.id===e.pointerId){lever.querySelector('.ce-knob').style.top=`${Math.max(0,Math.min(70,point(e).y*100-15))}%`;e.preventDefault();}});
      const release=(e,commit)=>{
        if(!drag||drag.id!==e.pointerId)return;const from=drag;drag=null;const p=point(e);
        lever.querySelector('.ce-knob').style.top='';if(lever.hasPointerCapture(e.pointerId))lever.releasePointerCapture(e.pointerId);
        if(commit&&p.x>=.1&&p.x<=.9&&p.y>=.72&&p.y<=1)act(lever.dataset.action,{x0:from.x,y0:from.y,x1:p.x,y1:p.y});
        e.preventDefault();
      };
      lever.addEventListener('pointerup',e=>release(e,true));lever.addEventListener('pointercancel',e=>release(e,false));
      lever.addEventListener('lostpointercapture',()=>{drag=null;lever.querySelector('.ce-knob').style.top='';});
    });
    draw();
  }
  window.WeirdCaptchaMechanics=window.WeirdCaptchaMechanics||{};
  window.WeirdCaptchaMechanics.comparator_engine={rootSelector:'.ce-engine',render};
})();

File: weird_captcha_gym/shared_runtime/server/incubator_graders/comparator_engine.py
"""Independent replay of primitive lever actions; never trust a final permutation."""
from __future__ import annotations
import math
MECHANIC_ID = 'comparator_engine'


def grade(payload, truth, public):
    def fail(message, outcome='invalid_transcript'):
        return {'graded': True, 'passed': False, 'score': 0, 'feedback': message, 'outcome': outcome}
    if any(not isinstance(x, dict) for x in (payload, truth, public)):
        return fail('malformed result')
    for key in ('mechanic_id', 'task_id', 'challenge_id'):
        if not truth.get(key) or any(x.get(key) != truth[key] for x in (payload, public)):
            return fail('stale task or challenge')
    if truth['mechanic_id'] != MECHANIC_ID:
        return fail('wrong mechanic')
    condition = truth.get('control_condition')
    if condition != public.get('control_condition') or payload.get('control_condition') != condition:
        return fail('wrong control condition')
    manual = truth.get('manual_readings', False)
    if type(manual) is not bool or public.get('manual_readings', False) is not manual or bool((condition or {}).get('difficulty_parameters', {}).get('manual_readings')) != manual:
        return fail('wrong comparison mode')
    mode = (condition or {}).get('interaction', 'simplified')
    source = {'simplified': 'button', 'full': 'lever_drag'}.get(mode)
    if source is None:
        return fail('unsupported interaction')
    try:
        if public['slides'] != truth['slides'] or public['runtime_weights'] != truth['weights'] or public['limits'] != truth['limits']:
            return fail('public/private contract mismatch')
        row = [s['id'] for s in truth['slides']]
        weights = truth['weights']
        limits = truth['limits']
        if not 4 <= len(row) <= 13 or len(set(row)) != len(row) or set(row) != set(weights):
            return fail('invalid slide bank')
        if any(type(v) is not int for v in weights.values()) or len(set(weights.values())) != len(row):
            return fail('invalid weights')
        if any(type(limits[k]) is not int or limits[k] < 1 for k in ('readings', 'levers')):
            return fail('invalid limits')
        if condition and condition.get('difficulty_parameters') != truth['parameters']:
            return fail('difficulty contract mismatch')
    except (KeyError, TypeError, ValueError):
        return fail('malformed comparator contract')
    events = payload.get('events')
    if not isinstance(events, list) or not 1 <= len(events) <= (2000 if manual else 400):
        return fail('missing or excessive actions')
    cursor = advances = exchanges = 0
    readings = 0 if manual else 1
    terminal = None
    for seq, event in enumerate(events, 1):
        if terminal is not None or not isinstance(event, dict) or type(event.get('seq')) is not int or event['seq'] != seq:
            return fail('invalid event sequence')
        if event.get('input_source') != source:
            return fail('wrong interaction input')
        if mode == 'full':
            gesture = event.get('gesture')
            if not isinstance(gesture, dict):
                return fail('missing lever geometry')
            try:
                x0,y0,x1,y1 = (gesture[k] for k in ('x0','y0','x1','y1'))
                if any(type(v) not in (int,float) or not math.isfinite(v) for v in (x0,y0,x1,y1)):
                    return fail('invalid lever geometry')
                if not (0.25 <= x0 <= .75 and 0 <= y0 <= .4 and .1 <= x1 <= .9 and .72 <= y1 <= 1):
                    return fail('lever did not reach the lower detent')
            except (KeyError, TypeError):
                return fail('invalid lever geometry')
        if type(event.get('cursor')) is not int or event['cursor'] != cursor or event.get('pair') != row[cursor:cursor+2]:
            return fail('stale carriage or slide identities')
        action = event.get('type')
        if action == 'seal':
            terminal = 'sorted' if all(weights[a] < weights[b] for a,b in zip(row,row[1:])) else 'unsorted_seal'
        elif action == 'weigh' and manual:
            if readings >= limits['readings']:
                terminal = 'comparison_exhausted'
            else:
                readings += 1
        elif action in ('advance', 'exchange'):
            if advances + exchanges >= limits['levers']:
                terminal = 'lever_exhausted'
            elif action == 'advance' and not manual and readings >= limits['readings']:
                terminal = 'comparison_exhausted'
            elif action == 'advance':
                cursor = (cursor + 1) % (len(row)-1)
                advances += 1
                readings += 0 if manual else 1
            else:
                row[cursor], row[cursor+1] = row[cursor+1], row[cursor]
                exchanges += 1
        else:
            return fail('unknown lever action')
    if terminal is None:
        return fail('frame has not been sealed')
    counts = payload.get('counts')
    if not isinstance(counts, dict) or set(counts) != {'advances', 'exchanges', 'readings'} or any(type(v) is not int for v in counts.values()):
        return fail('invalid action counts')
    if payload.get('final_order') != row or payload.get('counts') != {'advances': advances, 'exchanges': exchanges, 'readings': readings}:
        return fail('reported state disagrees with replay')
    passed = terminal == 'sorted'
    if payload.get('completed') is not passed:
        return fail('false completion claim')
    inversions = sum(weights[a] > weights[b] for i,a in enumerate(row) for b in row[i+1:])
    return {'graded': True, 'passed': passed, 'score': 100 if passed else 0, 'outcome': terminal,
            'feedback': 'Exact specimen order accepted' if passed else {
                'unsorted_seal': 'FAIL · Frame sealed out of order', 'lever_exhausted': 'FAIL · Lever reserve exhausted',
                'comparison_exhausted': 'FAIL · Comparison reserve exhausted'}[terminal],
            'metrics': {'advances': advances, 'exchanges': exchanges, 'readings': readings,
                        'lever_pulls': advances+exchanges, 'inversions': inversions}}
